bin/guard

```diff
@@ -2,31 +2,114 @@
 
 require "pathname"
 
-default = Pathname(ENV.fetch("BUNDLE_GEMFILE", "Gemfile"))
-relative_to_binstub = Pathname(__FILE__).realpath + "../../Gemfile"
-gemfile = [default, relative_to_binstub].detect(&:exist?)
-
-if gemfile && !ENV["RUBYGEMS_GEMDEPS"]
-  ENV["BUNDLE_GEMFILE"] = gemfile.to_s
-  require "rubygems"
-  require "bundler/setup"
-else
-  require "rubygems"
-end
+class GuardReloader
+  class Config
+    def using_rubygems?
+      ENV["RUBYGEMS_GEMDEPS"]
+    end
+
+    def setup_rubygems_for_deps
+      require "rubygems"
+    end
+
+    def current_bundler_gemfile
+      ENV["BUNDLE_GEMFILE"]
+    end
+
+    def using_bundler?
+      ENV["BUNDLE_GEMFILE"]
+    end
+
+    def setup_bundler_env(gemfile)
+      ENV["BUNDLE_GEMFILE"] = gemfile
+    end
+
+    def setup_bundler
+      require "rubygems"
+      require "bundler/setup"
+    end
+
+    def program_path
+      Pathname(__FILE__)
+    end
+
+    def program_arguments
+      ARGV
+    end
+
+    def windows?
+      Gem.win_platform?
+    end
+
+    def default_system_ruby
+      RbConfig.ruby
+    end
+
+    def exit_with(code)
+      exit(code)
+    end
+
+    def spawn_with(*args)
+      spawn(*args)
+    end
+
+    def wait_ignoring_interrupts(pid)
+      Process.wait2(pid)[1].exitstatus
+    rescue Interrupt
+      retry
+    rescue Errno::ECHILD
+      1
+    end
+
+    def exist?(path)
```

Comment:
The point is stubbing the config object as a whole - and having a guarantee that the main class contains only logic - and doesn't rely on anything else. Otherwise, setting up exist stubs is quite error-prone and especially when using a Pathname. (Though the Pathname usage here could be removed completely). "Reducing the number of methods" isn't a priority for code so sensitive to external factors and so hard to debug.